src/main/java/dan200/computercraft/core/terminal/TextBuffer.java

```diff
@@ -12,112 +12,40 @@
     public TextBuffer( char c, int length )
     {
         text = new char[length];
-        for( int i = 0; i < length; i++ )
-        {
-            text[i] = c;
-        }
+        this.fill( c );
     }
 
     public TextBuffer( String text )
     {
-        this( text, 1 );
-    }
-
-    public TextBuffer( String text, int repetitions )
-    {
-        int textLength = text.length();
-        this.text = new char[textLength * repetitions];
-        for( int i = 0; i < repetitions; i++ )
-        {
-            for( int j = 0; j < textLength; j++ )
-            {
-                this.text[j + i * textLength] = text.charAt( j );
-            }
-        }
-    }
-
-    public TextBuffer( TextBuffer text )
-    {
-        this( text, 1 );
-    }
-
-    public TextBuffer( TextBuffer text, int repetitions )
-    {
-        int textLength = text.length();
-        this.text = new char[textLength * repetitions];
-        for( int i = 0; i < repetitions; i++ )
-        {
-            for( int j = 0; j < textLength; j++ )
-            {
-                this.text[j + i * textLength] = text.charAt( j );
-            }
-        }
+        this.text = text.toCharArray();
     }
 
     public int length()
     {
         return text.length;
     }
 
-    public String read()
-    {
-        return read( 0, text.length );
-    }
-
-    public String read( int start )
-    {
-        return read( start, text.length );
-    }
-
-    public String read( int start, int end )
-    {
-        start = Math.max( start, 0 );
-        end = Math.min( end, text.length );
-        int textLength = Math.max( end - start, 0 );
-        return new String( text, start, textLength );
-    }
-
     public void write( String text )
     {
-        write( text, 0, text.length() );
+        write( text, 0 );
     }
 
     public void write( String text, int start )
     {
-        write( text, start, start + text.length() );
-    }
-
-    public void write( String text, int start, int end )
-    {
-        int pos = start;
         start = Math.max( start, 0 );
-        end = Math.min( end, pos + text.length() );
-        end = Math.min( end, this.text.length );
+        int end = Math.min( start + text.length(), this.text.length );
         for( int i = start; i < end; i++ )
         {
-            this.text[i] = text.charAt( i - pos );
+            this.text[i] = text.charAt( i - start );
```

Comment:
Sorry - should have spotted this earlier, but I don't believe this change is quite correct. Before, we had the following behaviour:

```java
buffer.write("Hello, world!", -3);
System.out.println(buffer);
// `lo, world!   '
```

Now we'd get:

```java
buffer.write("Hello, world!", -3);
System.out.println(buffer);
// `Hello, world!'
```

Reply:
[Admittedly I wasn't paying enough attention when updating tests](https://github.com/SquidDev-CC/CC-Tweaked/pull/738/commits/ddcc8ee854d68d24b0bd6a927bf0c49d57d6b75b#diff-f89c5b51d7e953968eb16197e4603fb6344906abb7459e0b16c40e27175e797bL107-R80), since I wrote the tests first to prevent such issues... will update later!

Reply:
979b1a8593e85589585291823e4698b76bd0a4f6

Code is the same as before now, just substituting `start + text.length()` for the `end` argument.
+ extra test for the example you gave